### ISSUE TITLE
Short Properties for immutable object simplification

### ISSUE THREAD

**Issue by sils1297** [coala-analyzer/coala#1006]
would be nice to have a specification for properties that simply make a variable read-only. example:

```
def __init__(self, match):
    self._match = match
    self.match = property(lambda self: self._match)
```

maybe we can make it a oneliner with a nifty function :)

**Comment by Makman2** [comment]
Ha that doesn't work!
```python
>>> class X:
...     def __init__(self, a):
...        self._a = a
...        self.a = property(lambda self: self._a)
...
>>> X(3).a
<property object at 0x7f7d6a7f57c8>
```
It places the property object itself into the field `a` ._.

**Comment by Makman2** [comment]
Ha that doesn't work!
```python
>>> class X:
...     def __init__(self, a):
...        self._a = a
...        self.a = property(lambda self: self._a)
...
>>> X(3).a
<property object at 0x7f7d6a7f57c8>
```
It places the property object itself into the field `a` ._.

**Comment by sils1297** [comment]
yeah I noticed that too, however there *was* a way of doing this *somehow*

**Comment by AbdealiJK** [comment]
The property needs to be assigned to the class and not to the object.
So, it should be `type(self)` or `self.__class__`

```
>>> class X:
...     def __init__(self, a):
...             self._a = a
...             self.__class__.a = property(lambda self: self._a)
... 
>>> X(3).a = 3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: can't set attribute
>>> X(3).a
3
```

I think the `@property` is more readable.

**Comment by Makman2** [comment]
I'm also more in favor of using the decorator and do everything manually. The only reason I would do this property assignment is when it has performance advantages, so this needs to be evaluated.

**Comment by AbdealiJK** [comment]
It would actually be the reverse I think. This method would be slower - as it gets assigned every time `__init__` gets called (whereas the decorator would only run once when the class is defined)

**Comment by Makman2** [comment]
Can't we put that outside of `__init__`, since this runs in class scope?

*Testing this...*

**Comment by Makman2** [comment]
Yes this works

**Comment by Makman2** [comment]
Do we still want that? I have no problems with out current design^^

**Comment by sils1297** [comment]
I'd really love to have a short version like this for readonly properties, that'd save a lot of hassle

**Comment by Makman2** [comment]
The coolest would be to use metaclasses, so it would work like this:
```python
class X:
    a = shortproperty
```
The metaclass would see that `a` is "annotated" as a shortproperty, and it would create the property inside `a` that fetches a value from `self._a`^^ But then we need to use the metaclass everytime on objects that want to use `shortproperty`... maybe there's a way to set a metaclass globally for all classes^^

**Comment by sils1297** [comment]
see also http://stackoverflow.com/questions/101268/hidden-features-of-python#102062

**Comment by AbdealiJK** [comment]
https://pypi.python.org/pypi/cached-property/1.2.0 is what i used in another project and it was very satisfactory

**Comment by sils1297** [comment]
I think this could be easily done with a descriptor, probably there's already one in the stdlib or so.
